File: libLogit.py
import json
import logging
from dataclasses import dataclass
from json import JSONDecodeError
from logging import Logger
from logging.handlers import SocketHandler
from pathlib import Path as _Path
from typing import Any, Dict, Iterable, Optional
from urllib.parse import urlparse

__all__ = ['init_from_config', 'LOG', 'ENDL', 'LogConfigurationError']

LEVEL_MAP = {
    'trace': logging.NOTSET,
    'debug': logging.DEBUG,
    'info': logging.INFO,
    'warn': logging.WARNING,
    'warning': logging.WARNING,
    'error': logging.ERROR,
    'fatal': logging.CRITICAL,
}

VALID_TOP_LEVEL_KEYS = {'level', 'timestamp', 'file_location', 'network_file_location'}
VALID_LEVEL_KEYS = {'threshold', 'tag'}


class LogConfigurationError(RuntimeError):
    """Raised when a configuration file fails validation."""


@dataclass
class _Config:
    threshold: str
    tag_level: bool
    timestamp: bool
    file_location: Optional[str]
    network_file_location: Optional[str]

    @classmethod
    def from_file(cls, path: str | _Path) -> '_Config':
        try:
            payload = json.loads(_Path(path).read_text(encoding='utf-8'))
        except FileNotFoundError as exc:
            raise LogConfigurationError(f'Configuration file not found: {path}') from exc
        except JSONDecodeError as exc:
            raise LogConfigurationError(f'Invalid JSON configuration: {exc}') from exc
        return cls.from_dict(payload)

    @classmethod
    def from_dict(cls, payload: Dict[str, Any]) -> '_Config':
        extra = set(payload.keys()) - VALID_TOP_LEVEL_KEYS
        if extra:
            raise LogConfigurationError(f'Unsupported configuration keys: {sorted(extra)}')

        level_cfg = payload.get('level')
        if level_cfg is None:
            raise LogConfigurationError('Missing required field: level')

        if isinstance(level_cfg, str):
            threshold = level_cfg
            tag_level = True
        elif isinstance(level_cfg, dict):
            extra_level = set(level_cfg.keys()) - VALID_LEVEL_KEYS
            if extra_level:
                raise LogConfigurationError(f'Unsupported level options: {sorted(extra_level)}')
            threshold = level_cfg.get('threshold')
            if threshold is None:
                raise LogConfigurationError('level.threshold is required when level is an object')
            tag_level = bool(level_cfg.get('tag', True))
        else:
            raise LogConfigurationError('level must be a string or object')

        normalized_threshold = _normalize_level_name(threshold)
        timestamp = bool(payload.get('timestamp', True))
        file_location = payload.get('file_location')
        network_location = payload.get('network_file_location')

        _validate_optional_path('file_location', file_location)
        _validate_optional_path('network_file_location', network_location)

        return cls(
            threshold=normalized_threshold,
            tag_level=tag_level,
            timestamp=timestamp,
            file_location=file_location or None,
            network_file_location=network_location or None,
        )

    @property
    def threshold_level(self) -> int:
        return LEVEL_MAP[self.threshold]


def _normalize_level_name(level: str) -> str:
    norm = level.strip().lower()
    if norm not in LEVEL_MAP:
        raise LogConfigurationError(f'Unknown level: {level}')
    if norm == 'warning':
        norm = 'warn'
    return norm


def _validate_optional_path(name: str, value: Any) -> None:
    if value is None:
        return
    if not isinstance(value, str):
        raise LogConfigurationError(f'{name} must be a string or null')
    if value.strip() == '':
        raise LogConfigurationError(f'{name} cannot be an empty string')


class _LoggerState:
    def __init__(self) -> None:
        self.config: Optional[_Config] = None
        self.logger: Optional[Logger] = None

    def configure(self, config: _Config) -> None:
        self.config = config
        logger = logging.getLogger('liblogit')
        logger.setLevel(config.threshold_level)
        logger.propagate = False

        for handler in list(logger.handlers):
            logger.removeHandler(handler)

        formatter = logging.Formatter(
            _build_format(config),
            datefmt='%Y-%m-%dT%H:%M:%S%z',
        )

        console_handler = logging.StreamHandler()
        console_handler.setFormatter(formatter)
        console_handler.setLevel(config.threshold_level)
        logger.addHandler(console_handler)

        for path in _filter_paths([config.file_location, config.network_file_location]):
            _Path(path).parent.mkdir(parents=True, exist_ok=True)
            file_handler = logging.FileHandler(path)
            file_handler.setFormatter(formatter)
            file_handler.setLevel(config.threshold_level)
            logger.addHandler(file_handler)

        if config.network_file_location and _looks_like_socket(config.network_file_location):
            parsed = urlparse(config.network_file_location)
            if parsed.hostname and parsed.port:
                socket_handler = SocketHandler(parsed.hostname, parsed.port)
                socket_handler.setFormatter(formatter)
                socket_handler.setLevel(config.threshold_level)
                logger.addHandler(socket_handler)

        self.logger = logger

    def log(self, level: str, message: str) -> None:
        if self.logger is None or self.config is None:
            raise RuntimeError('Logger not configured. Call init_from_config first.')
        normalized = _normalize_level_name(level)
        self.logger.log(LEVEL_MAP[normalized], message)


def _filter_paths(paths: Iterable[Optional[str]]) -> Iterable[str]:
    for path in paths:
        if path and not _looks_like_socket(path):
            yield path


def _looks_like_socket(target: str) -> bool:
    parsed = urlparse(target)
    return parsed.scheme in {'tcp', 'udp'} and parsed.hostname is not None and parsed.port is not None


def _build_format(config: _Config) -> str:
    parts: list[str] = []
    if config.timestamp:
        parts.append('%(asctime)s')
    if config.tag_level:
        parts.append('%(levelname)s')
    parts.append('%(message)s')
    return ' '.join(parts)


_STATE = _LoggerState()


def init_from_config(path: str | _Path) -> None:
    config = _Config.from_file(path)
    _STATE.configure(config)


class _LogBuilder:
    def __init__(self, level: str) -> None:
        self.level = _normalize_level_name(level)
        self.fragments: list[str] = []
        self._committed = False

    def __lshift__(self, value: Any) -> '_LogBuilder':
        if value is ENDL:
            return self.commit()
        self.fragments.append(_stringify(value))
        return self

    def commit(self) -> '_LogBuilder':
        if not self._committed and self.fragments:
            message = ''.join(self.fragments)
            _STATE.log(self.level, message)
            self._committed = True
        return self

    def __del__(self) -> None:
        if self.fragments and not self._committed:
            try:
                self.commit()
            except Exception:
                pass


ENDL = object()


def LOG(level: str) -> _LogBuilder:
    return _LogBuilder(level)


def _stringify(value: Any) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, (dict, list, tuple)):
        try:
            return json.dumps(value)
        except TypeError:
            pass
    return str(value)

File: languages/javascript/src/liblogit.js
"use strict";

const fs = require("fs");
const path = require("path");

const LEVELS = {
  trace: 0,
  debug: 1,
  info: 2,
  warn: 3,
  warning: 3,
  error: 4,
  fatal: 5,
};

const LEVEL_LABELS = {
  trace: "TRACE",
  debug: "DEBUG",
  info: "INFO",
  warn: "WARNING",
  error: "ERROR",
  fatal: "CRITICAL",
};

function normalizeLevel(level) {
  if (typeof level !== "string") {
    throw new TypeError(`level must be a string, got ${typeof level}`);
  }
  const normalized = level.trim().toLowerCase();
  if (!(normalized in LEVELS)) {
    throw new Error(`Unknown level: ${level}`);
  }
  return normalized === "warning" ? "warn" : normalized;
}

function normalizeSinks(value, localPath, remotePath) {
  if (value === undefined || value === null) {
    const sinks = ["console"];
    if (localPath) sinks.push("file");
    if (remotePath) sinks.push("network");
    return sinks;
  }
  if (!Array.isArray(value)) {
    throw new TypeError("sinks must be an array");
  }
  const seen = new Set();
  for (const sink of value) {
    if (!["console", "file", "network"].includes(sink)) {
      throw new Error(`Unknown sink: ${sink}`);
    }
    seen.add(sink);
  }
  return Array.from(seen);
}

class LOGIT {
  constructor(structure = {}) {
    if (structure === null || typeof structure !== "object" || Array.isArray(structure)) {
      throw new TypeError("LOGIT structure must be an object");
    }
    this.name = structure.name || "default";
    this._level = normalizeLevel(structure.level || "info");
    this._localPath = structure.localPath || structure.local_path || structure.path || structure.file_location || null;
    this._remotePath =
      structure.remotePath ||
      structure.remote_path ||
      structure.network_path ||
      structure.network_file_location ||
      null;
    this.timestamp = structure.timestamp !== undefined ? Boolean(structure.timestamp) : true;
    this.tagLevel =
      structure.tagLevel !== undefined
        ? Boolean(structure.tagLevel)
        : structure.tag_level !== undefined
          ? Boolean(structure.tag_level)
          : true;
    this.format = structure.format || "text";
    this.metadata = structure.metadata || {};
    this.enabled = structure.enabled !== undefined ? Boolean(structure.enabled) : true;
    this._sinks = normalizeSinks(structure.sinks, this._localPath, this._remotePath);
  }

  get level() {
    return this._level;
  }

  set level(value) {
    this._level = normalizeLevel(value);
  }

  get localPath() {
    return this._localPath;
  }

  set localPath(value) {
    this._localPath = value ? String(value) : null;
    if (this._localPath && !this._sinks.includes("file")) {
      this._sinks.push("file");
    }
    if (!this._localPath) {
      this._sinks = this._sinks.filter((sink) => sink !== "file");
    }
  }

  get path() {
    return this.localPath;
  }

  set path(value) {
    this.localPath = value;
  }

  get remotePath() {
    return this._remotePath;
  }

  set remotePath(value) {
    this._remotePath = value ? String(value) : null;
    if (this._remotePath && !this._sinks.includes("network")) {
      this._sinks.push("network");
    }
    if (!this._remotePath) {
      this._sinks = this._sinks.filter((sink) => sink !== "network");
    }
  }

  get sinks() {
    return this._sinks.slice();
  }

  set sinks(value) {
    this._sinks = normalizeSinks(value, this._localPath, this._remotePath);
  }

  at(level) {
    return new LogBuilder(this, level);
  }

  log(level, message) {
    const normalized = normalizeLevel(level);
    if (!this.enabled || LEVELS[normalized] < LEVELS[this._level]) {
      return;
    }
    const rendered = this.render(normalized, stringify(message));
    if (this._sinks.includes("console")) {
      process.stderr.write(`${rendered}\n`);
    }
    if (this._sinks.includes("file") && this._localPath) {
      appendLine(this._localPath, rendered);
    }
    if (this._sinks.includes("network") && this._remotePath && !looksLikeSocket(this._remotePath)) {
      appendLine(this._remotePath, rendered);
    }
  }

  render(level, message) {
    if (this.format === "json") {
      const event = {
        level,
        logger: this.name,
        message,
      };
      if (Object.keys(this.metadata).length > 0) {
        event.metadata = this.metadata;
      }
      if (this.timestamp) {
        event.timestamp = new Date().toISOString();
      }
      return JSON.stringify(sortObject(event));
    }

    const parts = [];
    if (this.timestamp) {
      parts.push(new Date().toISOString());
    }
    if (this.tagLevel) {
      parts.push(LEVEL_LABELS[level]);
    }
    parts.push(message);
    return parts.join(" ");
  }
}

class LogBuilder {
  constructor(logit, level) {
    this.logit = logit;
    this.level = level;
    this.fragments = [];
    this.committed = false;
  }

  append(value) {
    if (!this.committed) {
      this.fragments.push(stringify(value));
    }
    return this;
  }

  commit() {
    if (!this.committed && this.fragments.length > 0) {
      this.committed = true;
      this.logit.log(this.level, this.fragments.join(""));
    }
    return this;
  }
}

function loadLogits(configPath) {
  const payload = JSON.parse(fs.readFileSync(configPath, "utf8"));
  if (!payload.logits) {
    return {
      default: new LOGIT({
        name: "default",
        level: typeof payload.level === "string" ? payload.level : payload.level.threshold,
        tag_level: typeof payload.level === "object" ? payload.level.tag : true,
        timestamp: payload.timestamp,
        path: payload.file_location,
        remotePath: payload.network_file_location,
      }),
    };
  }

  const defaults = payload.defaults || {};
  const result = {};
  for (const [name, config] of Object.entries(payload.logits)) {
    result[name] = new LOGIT({ ...defaults, ...config, name: config.name || name });
  }
  return result;
}

function appendLine(target, line) {
  const parent = path.dirname(target);
  if (parent && parent !== ".") {
    fs.mkdirSync(parent, { recursive: true });
  }
  fs.appendFileSync(target, `${line}\n`, "utf8");
}

function looksLikeSocket(target) {
  return /^tcp:\/\/|^udp:\/\//i.test(target);
}

function stringify(value) {
  if (typeof value === "string") {
    return value;
  }
  if (value === null || value === undefined) {
    return String(value);
  }
  if (typeof value === "object") {
    return JSON.stringify(value);
  }
  return String(value);
}

function sortObject(value) {
  return Object.keys(value)
    .sort()
    .reduce((result, key) => {
      result[key] = value[key];
      return result;
    }, {});
}

module.exports = {
  LOGIT,
  Logit: LOGIT,
  loadLogits,
  levels: {
    TRACE: "trace",
    DEBUG: "debug",
    INFO: "info",
    WARN: "warn",
    ERROR: "error",
    FATAL: "fatal",
  },
};
